MemoryReader.

wouterboomsma opened the pull request:
MemoryReader: Support for loading MDAnalysis trajectories into a numpy array, interfaced through the standard MDAnalysis reader interface. Makes it possible to manipulate coordinates in memory - for instance aligning frames in-memory.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [ ] CHANGELOG updated?
 -  Issue raised/referenced? — n/a (see PR #797 for context)

jbarnoud commented on package/MDAnalysis/coordinates/base.py:
```diff
@@ -121,6 +121,7 @@
 from six.moves import range
 import six
 
+import logging as log
```
Is this used anywhere?

wouterboomsma replied:
Nope. Removed in 5710477.

jbarnoud commented:
I will be quite busy until the end of next week. I'll review the PR arount the 19th if nobody did it until then.

jbarnoud commented on package/MDAnalysis/coordinates/DCD.py:
```diff
@@ -500,7 +500,7 @@ def _read_frame(self, frame):
         return ts
 
-    def timeseries(self, asel, start=None, stop=None, step=None, skip=None,
+    def timeseries(self, asel=None, start=None, stop=None, step=None, skip=None,
```
`asel=None` should be documented

wouterboomsma replied:
Good point. Fixed in 5710477.

orbeckst commented on package/MDAnalysis/coordinates/memory.py:
```diff
@@ -0,0 +1,255 @@
+:Maintainer: Wouter Boomsma <wb@bio.ku.dk>, wouterboomsma on github
+
+
+.. versionadded:: 0.15.0
```
Going to be 0.16.0

wouterboomsma replied:
Strange that I missed that one. Fixed in 2637a20.

orbeckst commented on package/MDAnalysis/coordinates/memory.py:
```diff
@@ -0,0 +1,255 @@
+
+    from MDAnalysis import Universe
+    from MDAnalysisTests.datafiles import DCD, PDB_small
+    from MDAnalysis.coordinates.array import ArrayReader
```
update docs: 
* coordinates.array --> coordinates.memory
* ArrayReader --> MemoryReader

wouterboomsma replied:
Sorry about these leftovers from the previous names. I should have checked it more thoroughly. Fixed in 2637a20.

orbeckst commented on package/MDAnalysis/coordinates/memory.py:
```diff
@@ -0,0 +1,255 @@
+
+    format = 'memory'
+
+    class MemoryTimestep(base.Timestep):
```
For the other classes we define `Timestep` at the module level. Follow that pattern for consistency (so you can just name it `Timestep` instead of `MemoryTimestep`)

wouterboomsma replied:
Fixed in 2637a20.

orbeckst commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -5028,6 +5037,46 @@ def _matches_unpickling(self, anchor_name, n_atoms, fname, trajname):
+
+        from ..coordinates.memory import MemoryReader
+
+        if self.trajectory.format != "array":
```
Are you using "memory" or "array" as the format name? `MemoryReader.format == "memory" ?

wouterboomsma replied:
Yes, I'm should be using "memory". Another embarrassing leftover. Sorry. Fixed in 2637a20.

orbeckst commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -5028,6 +5037,46 @@ def _matches_unpickling(self, anchor_name, n_atoms, fname, trajname):
+            except AttributeError:
+                coordinates = \
+                    np.array([ts.positions for ts in
+                            self.trajectory[frame_interval-1::frame_interval]])
```
Does `self.trajectory[frame_interval-1::frame_interval]` mimic some idiosyncratic behavior of `timeseries`? Otherwise I would have expected  `self.trajectory[::frame_interval]`.

wouterboomsma replied:
I fixed this in the the MemoryReader itself, but you are right that it should be fixed here as well. Also updated the corresponding test. Fixed in 2637a20.

orbeckst commented:
I ticked tests and docs (although see comments).

richardjgowers commented on package/MDAnalysis/coordinates/memory.py:
```diff
@@ -0,0 +1,255 @@
+
+    """
+
+    format = 'memory'
```
`format` here must be upper case

wouterboomsma replied:
I followed the style used by other readers. The `DCDReader` class for instance has a `format` class variable. However, I'd be happy to change it if there is a consensus about this - just let me know.

orbeckst replied:
@richardjgowers – when I look through the readers/writers I only see lower case `format`. 
```
grep -e 'Format =' MDAnalysis/coordinates/*py
```

Or do you mean, the format name should be `"Memory"`?

richardjgowers replied:
Right sorry, I meant `format = 'MEMORY'`.  All the guessers etc currently cast all format strings to uppercase.

wouterboomsma replied:
Ah. That makes sense. Fixed in 2b30138.

richardjgowers commented on package/MDAnalysis/coordinates/memory.py:
```diff
@@ -0,0 +1,255 @@
+            coordinates)
+        """
+        self.coordinate_array = coordinate_array
+        self.format = format
```
`self.format` here will override the `class.format` you've defined above.  This name needs to get changed to something else

wouterboomsma replied:
Well spotted. Although this never leads to problems in practice (the class variable is only read on uninstantiated classes, and the object variables only on instances), it is extremely confusing and ugly. I've renamed `self.format` to `self.stored_format`. Fixed in 2637a20.

richardjgowers commented on package/MDAnalysis/coordinates/memory.py:
```diff
@@ -0,0 +1,255 @@
+            array = np.swapaxes(array, 0, 2)
+        elif format[2] == self.format[2]:
+            array = np.swapaxes(array, 0, 1)
+        elif self.format[1] == format[0]:
```
for consistency, can this safely get changed to `format[0] == self.format[1]` (or whatever `self.format` gets renamed to

wouterboomsma replied:
Fixed in 2637a20.

richardjgowers commented on package/MDAnalysis/core/Selection.py:
```diff
@@ -178,6 +178,12 @@ def __init__(self, parser, tokens):
+        # in the corresponding universe, in which case this
+        # is returned directly. This works since the Universe.atoms
+        # are unique by construction.
+        if group.universe and group is group.universe.atoms:
```
This probably isn't doing what you think it is.

```python
if True and False is False:
    print 'This gets printed'
```
Because the implicit brackets are `(if True) and (False is False)`.

Doesn't just `if group is group.universe.atoms` work?

wouterboomsma replied:
Actually, it is doing what I think it is :), although I don't think it is necessary. I test for whether `group.universe` is different from `None` before reading `group.universe.atoms` just to be sure I don't get an `AttributeError`. As far as I could see, group.universe is always set, however, I became uncertain after seeing this definition:
```python
    @property
    def dimensions(self):
        """Dimensions of the Universe to which the group belongs, at the current time step."""
        if self.universe is not None:
            return self.universe.dimensions
        else:
            raise AttributeError("This AtomGroup does not belong to a Universe with a dimension.")
```
Therefore, I added this extra check (although I agree it should perhaps have been made more explicit by writing `group.universe is not None`. I would be happy to change it to what you suggest, if universe is indeed always set.

orbeckst replied:
I also think that @wouterboomsma 's code does what it is supposed to do.

I admit that I don't remember if we can really have AtomGroups that don't belong to a Universe. I don't think that this should happen under normal circumstances. In any case, for the future (I.e., #363) I defer to @richardjgowers and @dotsdl – they should know if this will be an issue for the future.

richardjgowers replied:
Yeah moving forward **everything** will have a Universe

@wouterboomsma I'd just change it to `if group is group.universe.atoms`

wouterboomsma replied:
Fixed in b503176.

wouterboomsma commented:
@orbeckst You expressed some interest in the `merge_universes` function in the `utils.py` file of encore (PR #797). Would it make sense to make this function part of this PR, perhaps adding it somewhere under `MDAnalysis.lib`? If so, where?
Here it is, for reference:
```python
def merge_universes(universes):
    """
    Merge list of universes into one

    Parameters
    ----------
    `universes` : list of Universe objects


    Returns
    ----------
    Universe object
    """

    for universe in universes:
        universe.transfer_to_memory()

    return mda.Universe(
        universes[0].filename,
        np.concatenate(tuple([e.trajectory.timeseries() for e in universes]),
        axis=1),
        format=MemoryReader)
```

orbeckst commented:
Re: `merge_universes()`: This is an interesting function that should be more generally exposed but `lib` is not the right place because anything under `lib` should be independent of the core MDAnalysis machinery such as AtomGroup, Universe, etc... At the moment it would fit into `AtomGroup.py` (similar to the `Merge()` function) but because the #363 rewrite will completely re-organize `AtomGroup.py` and friends I suggest we just hold of with moving `merge_universes()` elsewhere and wait until the dust around PR #815 settles. Maybe we find a good spot for "topology manipulation".

On the plus side for this PR: less work for you :-).

richardjgowers commented:
@wouterboomsma looking good once last comments are done

kain88-de commented:
@wouterboomsma did you ever test how big the speed improvement is for iterative algorithms using the memory reader? Would be nice to know for annoucement news of the new version.

wouterboomsma commented:
@richardjgowers, I ended up replacing `trajectory.format != 'memory'` with `trajectory.format != MemoryReader.format`, to avoid having the raw string specified twice (I replaced "memory" with "MEMORY" in the MemoryReader itself.

wouterboomsma commented:
@kain88-de I don't really have a good enough overview over the main bottlenecks in MDAnalysis to know which speed tests would be meaningful to do. Are you thinking of anything in particular?

richardjgowers commented:
@wouterboomsma yeah that's neater, though thinking about it though, `if isinstance(self.trajectory, MemoryReader)` will be most future-proof

wouterboomsma commented:
@richardjgowers I agree completely. Fixed in f4e3b65.

richardjgowers commented:
@wouterboomsma sorry about the double change, late night code reviews!

WRT speed changes, I think we're thinking of advertising it in a blog post, so something basic like

```python
from MDAnalysis.tests.datafiles import PSF, DCD  # use the reference system

u = mda.Universe(PSF, DCD)

ag = u.select_atoms('resid 1')
for ts in u.trajectory:
    print ag.center_of_mass()
```

To see if this gets much faster if the trajectory is transferred to memory (and the setup cost of transferring into memory).

There's a few calculations I can think of (eg mean squared displacement) which require a lot of thrashing around the trajectory, so it'd be cool if the MemoryReader made this faster.  A quick iteration test should tell if this is the case.

kain88-de commented:
> mean squared displacement

That is a good example since it requires reading the trajectory a lot. Or just running a `Contacts`/`Align` analysis that we provide.

wouterboomsma commented:
@richardjgowers, @kain88-de Generally, I wouldn't expect large improvements as long as you formulate it as a iterative problem. The benefits would arise by letting numpy do the entire calculation at once - but this means you cannot use the calculations in AtomGroup directly, at least as far as I can see. Here is an extrapolation of the example you sent, where the first two versions use the standard iterative approach with and without in_memory, and the last uses a pure numpy approach:
```python
import time
import numpy as np
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import PSF, DCD

t0 = time.time()
u1 = mda.Universe(PSF, DCD)
t1 = time.time()
print "Read universe from DCD: {0}".format(t1-t0)

u2 = mda.Universe(PSF, DCD)

t0 = time.time()
u2.transfer_to_memory()
t1 = time.time()
print "Transfer_to_memory: {0}".format(t1-t0)

ag = u1.select_atoms('resid 1')
coms1 = []
t0 = time.time()
for ts in u1.trajectory:
    coms1.append(ag.center_of_mass())
t1 = time.time()
print "Iterative procedure: {0}".format(t1-t0)

ag2 = u2.select_atoms('resid 1')
coms2 = []
t0 = time.time()
for ts in u2.trajectory:
    coms2.append(ag2.center_of_mass())
t1 = time.time()
print "Iterative procedure in memory: {0}".format(t1-t0)

t0 = time.time()
coms3 = np.average(u2.trajectory.timeseries(ag2), weights=(ag2.masses/np.sum(ag2.masses)), axis=0)
t1 = time.time()
print "Non-iterative procedure: {0}".format(t1-t0)

# Verify that solutions are the same
assert((abs(np.array(coms1) - np.array(coms2)) < 0.0001).all())
assert((abs(np.array(coms1) - coms3) < 0.0001).all())
```
The timings I get are:
```
Read universe from DCD: 0.120114088058
Transfer_to_memory: 0.0131919384003
Iterative procedure: 0.00638604164124
Iterative procedure in memory: 0.00462794303894
Non-iterative procedure: 0.00023889541626
```
Several things are clear. The first read dominates any calculations that we do afterwards, so this is probably not the best test. Other than that, we can see that the numpy calculation is about 20 times as fast - but only if you execute it as a non-iterative procedure. I don't think the difference between "iterative procedure" and "iterative procedure in memory" are significant (they fluctuate around similar values).

wouterboomsma commented:
Btw, I got this strange TimeOutError problem again, leading to a failed Travis build. This is probably caused by the fact that `test_atomgroup.py` has been expanded a bit. I therefore just upped the travis process timeout to 400s.

kain88-de commented:
@wouterboomsma DCD is a bit unfair since it is almost just a memory map. XTC are PDB should give a fairer comparison since they take longer to read. That would also be good information for people who use Gromacs.

Still I'm a bit disappointed that the in memory approach doesn't help that much in the iterative approach. I hoped for better values even with DCD. This is not the fault of your memory reader though, it's a general problem of the iterations.

wouterboomsma commented:
@kain88-de 

Here is an updated version, now using XTCReader, and repeating each of the main tests a thousand time to get some more reliable numbers:
```python
import time
import numpy as np
import MDAnalysis.analysis.align as align
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import PSF, DCD, GRO, XTC

# Setup universes
t0 = time.time()
u1 = mda.Universe(GRO, XTC)
t1 = time.time()
print "Read universe: {0}".format(t1-t0)
u2 = mda.Universe(GRO, XTC)

# Measure time for transfer to memory
t0 = time.time()
u2.transfer_to_memory()
t1 = time.time()
print "Transfer_to_memory: {0}".format(t1-t0)

# Standard case: iterative procedure
ag = u1.select_atoms('resid 1')
t0 = time.time()
for i in range(1000):
    coms1 = []
    for ts in u1.trajectory:
        coms1.append(ag.center_of_mass())
t1 = time.time()
print "Iterative procedure: {0}".format(t1-t0)

# Iterative procedure in-memory
ag2 = u2.select_atoms('resid 1')
t0 = time.time()
for i in range(1000):
    coms2 = []
    for ts in u2.trajectory:
        coms2.append(ag2.center_of_mass())
t1 = time.time()
print "Iterative procedure in memory: {0}".format(t1-t0)

# Direct numpy calculation in-memory
t0 = time.time()
for i in range(1000):
    coms3 = np.average(u2.trajectory.timeseries(ag2), weights=(ag2.masses/np.sum(ag2.masses)), axis=0)
t1 = time.time()
print "Non-iterative procedure: {0}".format(t1-t0)

# Direct numpy calculation without overhead of timeseries() call
array = u2.trajectory.timeseries(ag2)
t0 = time.time()
for i in range(1000):
    coms3 = np.average(array, weights=(ag2.masses/np.sum(ag2.masses)), axis=0)
t1 = time.time()
print "Non-iterative procedure excluding timeseries() call: {0}".format(t1-t0)

# Verify solutions are the same
assert((abs(np.array(coms1) - np.array(coms2)) < 0.0001).all())
assert((abs(np.array(coms1) - coms3) < 0.0001).all())
```
Times on my laptop are now:
```
Read universe: 0.62521314621
Transfer_to_memory: 0.0403649806976
Iterative procedure: 36.6376540661
Iterative procedure in memory: 0.486598968506
Non-iterative procedure: 6.78375792503
Non-iterative procedure excluding timeseries() call: 0.0540280342102
```
Indeed, we here see some interesting differences. The in-memory version is about 75 times faster than the standard XTCReader. Surprisingly, in this case, it is also faster than the non-iterative numpy procedure. It turns out that this is due to the overhead in the timeseries() call. The last value shows that there is another potential factor of 10 to be gained by extracting the numpy array once and for all (or if we removed some of the overhead in the timeseries() call)

kain88-de commented:
Cool that's more like it.

richardjgowers commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -652,15 +653,18 @@ def rms_fit_trj(
          - ``False``: show all status messages and do not change the the logging
            level (default)
-
+      *in_memory*
```
Maybe mention that this is faster?

wouterboomsma replied:
I haven't tested the speed very systematically for the alignment, so I've just added ", which can improve performance substantially in some cases".

richardjgowers commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -682,19 +686,24 @@ def rms_fit_trj(
+    writer = None
+    if in_memory:
+        traj.transfer_to_memory()
+        frames = traj.trajectory
```
You can just move the above `frames =` to below this if block to simplify

wouterboomsma replied:
Do you mean moving the `frames = traj.trajectory` line from the first line of the method (line 684) to after the if-block (line 709). If so, this would not work, since the else branch in the if-statement also uses the `frames` variable. Perhaps I'm misunderstanding what you meant.

richardjgowers replied:
Ah my mistake, didn't see that

orbeckst commented:
CHANGELOG entry and done?

wouterboomsma commented:
@richardjgowers Regarding your tangent, do you mean implementing a `MemoryWriter`? For completeness, it would make sense - but do you think it is a common use case? Regarding the hdf5 format, that would also be nice to have - but wouldn't it be most natural to have this functionality in a dedicated `HDF5Reader` and `HDF5Write`r?

wouterboomsma commented:
Regarding the CHANGELOG, there seem to be no entries for 0.16.0 yet, and I'm not entirely sure about the format, but here is a suggestion:

Enhancements

  * Added 'MemoryReader' class to allow manipulation of trajectory data in-memory, which can provide substantial speed-ups to certain calculations. The constructor of 'Universe' now takes an 'in_memory' option, which will transfer the corresponding trajectory to memory. Likewise a new 'Universe.transfer_to_memory()' method makes it possible to make this transfer after construction of the universe. 
  * Added 'in_memory' option to 'rms_fit_trj', which makes it possible to do in-place (in-memory) alignment of trajectories.

richardjgowers commented:
@wouterboomsma thanks, took care of the CHANGELOG for you

orbeckst commented:
Put it under the 0.15.1 heading in your CHANGELOG. Once you merge/rebase against develop you will see that the 0.15.1 was changed to 0.16.0

Keep your entries shorter, put them under Enhancements, and reference the PR (or issue) number. Don't forget to add authors' handles to the entry and if they don't appear yet in AUTHORS also add everyone to that file. 

--
Oliver Beckstein
email: orbeckst@gmail.com


> Am Sep 15, 2016 um 7:41 schrieb Wouter Boomsma <notifications@github.com>:
> 
> Regarding the CHANGELOG, there seem to be no entries for 0.16.0 yet, and I'm not entirely sure about the format, but here is a suggestion:
> 
> Enhancements
> 
> Added 'MemoryReader' class to allow manipulation of trajectory data in-memory, which can provide substantial speed-ups to certain calculations. The constructor of 'Universe' now takes an 'in_memory' option, which will transfer the corresponding trajectory to memory. Likewise a new 'Universe.transfer_to_memory()' method makes it possible to make this transfer after construction of the universe.
> Added 'in_memory' option to 'rms_fit_trj', which makes it possible to do in-place (in-memory) alignment of trajectories.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

orbeckst commented:
Hooray, congratulations and thanks a lot!

--
Oliver Beckstein
email: orbeckst@gmail.com


> Am Sep 15, 2016 um 8:05 schrieb Richard Gowers <notifications@github.com>:
> 
> Merged #976.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

wouterboomsma commented:
Great! Thank you all for your time in reviewing this PR.
